### PR TITLE
Allow Trace Samplers to Modify the Activity Trace State

### LINKED ISSUE #63652: Samplers should be allowed to modify tracestate
https://github.com/open-telemetry/opentelemetry-specification/pull/988
https://github.com/open-telemetry/opentelemetry-dotnet/issues/1708

### PR DESCRIPTION
**Pull request by tarekgh**
Fixes https://github.com/dotnet/runtime/issues/63652

### CODE REVIEW

**Comment by tarekgh** [comment]
@noahfalk could you please help reviewing this? 

CC @ericstj for the `System.Runtime.CompilerServices.IsExternalInit` type forward.

CC @cijothomas @reyang

File: src/libraries/System.Diagnostics.DiagnosticSource/tests/ActivitySourceTests.cs [hunk 1]
```diff
@@ -1073,6 +1073,147 @@ public void TestAddingMorePropertiesWithActivityCreate()
...
+                };
+                listener.SampleUsingParentId = sampleUsingParentId;
+
+                bool setTraceWithContextParent = false;
+                string? traceStateValueWithContextParent = null;
+
+                SampleActivity<ActivityContext> sampleUsingContext = (ref ActivityCreationOptions<ActivityContext> activityOptions) => {
+                    if (setTraceWithContextParent)
+                    {
+                        activityOptions = activityOptions with { TraceState = traceStateValueWithContextParent };
```
Comment (reyang) [inline src/libraries/System.Diagnostics.DiagnosticSource/tests/ActivitySourceTests.cs:1107]: Just to make sure we are on the same page:

We've discussed about this C# 10 language feature and the understanding is: if we are compiling against an old version of runtime which doesn't provide full support for all C# 10 features, we would get compile time error if such C# 10 feature is not going to work on the target runtime. In other words, if we don't get compile time error, then we have confidence that such feature would work fine (not having runtime error) on the platforms we targeted.

Comment (reyang) [reply src/libraries/System.Diagnostics.DiagnosticSource/tests/ActivitySourceTests.cs:1107]: Regarding why I would call this out - there are some concerns/questions which is beyond my knowledge https://github.com/open-telemetry/opentelemetry-dotnet/discussions/2790#discussion-3800306.

Comment (tarekgh) [reply src/libraries/System.Diagnostics.DiagnosticSource/tests/ActivitySourceTests.cs:1107]: > Regarding why I would call this out - there are some concerns/questions which are beyond my knowledge https://github.com/open-telemetry/opentelemetry-dotnet/discussions/2790#discussion-3800306.

I replied there too. I will be happy to help with any questions raised there too.

**Review by reyang — APPROVED** [review]
LGTM.

**Review by noahfalk — APPROVED** [review]
LGTM : )

File: src/libraries/System.Diagnostics.DiagnosticSource/tests/ActivitySourceTests.cs [hunk 2]
```diff
@@ -1073,6 +1073,147 @@ public void TestAddingMorePropertiesWithActivityCreate()
...
+                ActivitySource.AddActivityListener(listener);
+
+                bool setTraceWithStringParent = false;
+                string? traceStateValueWithStringParent = null;
+
+                SampleActivity<string> sampleUsingParentId = (ref ActivityCreationOptions<string> activityOptions) =>
+                {
+                    if (setTraceWithStringParent)
+                    {
+                        activityOptions = activityOptions with { TraceState = traceStateValueWithStringParent };
```
[the remark this answers is not in the document]
Comment (tarekgh) [reply src/libraries/System.Diagnostics.DiagnosticSource/tests/ActivitySourceTests.cs:1096]: No, to use this feature must use C# compiler version 10 or later. This is not the first feature we expose in the framework using this new syntax. [System.Text.Json](https://github.com/dotnet/runtime/blob/ae259bff32bd7fb5eee4d5a86a42214c39aba362/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonObjectInfoValuesOfT.cs#L20) already using it a lot.

[the remark this answers is not in the document]
Comment (cijothomas) [reply src/libraries/System.Diagnostics.DiagnosticSource/tests/ActivitySourceTests.cs:1096]: yes. OpenTelemetry SDK uses 10

**Review by ericstj — COMMENTED** [review]
LGTM

**Pull request merged**